new(sysdig): add support for falco-compatible plugin configuration file

LucaGuerra opened the pull request:
See https://falco.org/docs/configuration/ and https://falco.org/docs/plugins/#loading-plugins-in-falco to understand how this works. The PR is still WIP because it needs to be tested and debugged but I wanted to show the approach.
Up until now, the general idea was that:
* If you use `-H` you will tell Sysdig to init the specified plugins with init config
* With `-I` you can specify the input plugin to be loaded and its open params
* If you add `-I` and no `-H` it will init all plugins from the configured directories and then do as above
Now you can also: use `--plugin-config-file` to supply a falco compatible config file that can specify the plugins paths and arguments.

Note that this is not compatible with Windows and MacOS yet because of the json/yaml dependency and a bit of a mess in cmake.

Commit:
Add support for falco-compatible plugin configuration file

Signed-off-by: Luca Guerra <luca.guerra@sysdig.com>

FedeDP commented on userspace/sysdig/plugin_utils.h:
```diff
@@ -38,14 +40,70 @@ typedef struct plugin_dir_info
\ No newline at end of file
+// Select a plugin for initialization. Name can be either a name or a path.
+void select_plugin_init(string& name, const string& init_config);
+// Select a plugin for use. In case of a source plugin, that the opened one.
```
Typo on this comment?

LucaGuerra replied:
No, it's an example of modern art poetry. 😬 The artist wanted to convey the deep meaning of openness.

FedeDP reviewed (commented):
Left a very minor comment. 
It LGTM but i reserve to give it another look because it is quite hard to grasp altogether!

LucaGuerra commented:
Checked if it compiles on different platforms too. It does work with dependencies but I didn't test plugins on those yet.
* MacOS 
* Windows with VS2022 MSBuild CLI (if this patch https://github.com/falcosecurity/libs/pull/255 is present)

LucaGuerra commented:
>  It LGTM but i reserve to give it another look because it is quite hard to grasp altogether!

The important thing is to try it out / have ideas to try and see if it works in all the ways we can open plugins.

Commit:
fix(cmake): bundled deps guard only for yaml-cpp

Signed-off-by: Luca Guerra <luca.guerra@sysdig.com>

FedeDP committed:
fix(userspace/sysdig): properly map a plugin to user-passed name, as it can also be a path to the plugin shared object.

This fixes usages like:
`sysdig -H ./libdummy.so -I ./libdummy.so`

Signed-off-by: Federico Di Pierro <nierro92@gmail.com>

FedeDP commented:
I committed a small fix: we need to maintain a map between plugin and user-specified plugin-name, because we support specifying a path to a plugin too (not only a plugin name).

FedeDP commented on userspace/sysdig/plugin_utils.cpp:
```diff
@@ -155,39 +163,146 @@ void select_plugin(string& name, const string& init_config)
-	}
-	auto plugin = sinsp_plugin::register_plugin(inspector, itr->second.path, itr->second.init_config.c_str());
-	return plugin;
+    g_selected_plugins_enable.emplace(name, open_params);
```
@LucaGuerra btw why are we allowing to enable multiple source plugin just to raise an exception (below in `enable_source_plugin`) if more than one is present?

leogr reviewed (commented):
Hey @LucaGuerra 

Overall, this PR SGMT.  I will give it a send look soon.

Meanwhile, although it's not strictly related to this PR, I have just noticed that plugins are loaded from `/usr/local/share/sysdig/plugins/` but the command usage help says `/usr/share/sysdig/plugins` (and my `SYSDIG_PLUGIN_DIR` env var is empty).

Also, I did not find the following text very clear:
```
 -I <inputname>[:<openparams>], --input <inputname>[:<openparams>]                                                                                                                                        
                    registers a previously selected plugin.                                                                                                                                               
                    If the plugin is a source plugin, capture events using it, passing to the                                                                                                             
                    plugin the openparams string as open parameters.                                                                                                                                      
                    Only a single source plugin can be registered.                                                                                                                                        
                    If no plugins were selected, registers any found plugin in the directories                                                                                                            
                    specified by ;-separated environment variable SYSDIG_PLUGIN_DIR and                                                                                                                   
                    in /usr/share/sysdig/plugins; then use the provided one as input source.
```

I'd expect `<pluginname>` instead of `<inputname>`. Also `--input` can be very misleading in case of a non-source plugin.

FedeDP committed:
fix(userspace/sysdig): fix sysdig helper strings for chisels and plugins system folders.

Signed-off-by: Federico Di Pierro <nierro92@gmail.com>

FedeDP committed:
update(userspace/sysdig): inputname -> pluginname.

Signed-off-by: Federico Di Pierro <nierro92@gmail.com>

FedeDP commented:
@leogr should've addressed your issues!
> Also --input can be very misleading in case of a non-source plugin.

Yeah i agree; i am not sure if we can change it without breaking backward compatibility...

FedeDP committed:
update(userspace/sysdig): fixed up plugins support.

Updated helper text; dropped useless g_selected_plugins_init map. Fixed logic.

Signed-off-by: Federico Di Pierro <nierro92@gmail.com>

FedeDP commented:
@leogr  imrpoved the helper text with the actual behavior ('-I' always specifies the input plugin to be used!)

leogr commented on userspace/sysdig/sysdig.cpp:
```diff
@@ -186,29 +186,29 @@ static void usage()
+"                    openparams string as open parameters.\n"
 "                    Only a single source plugin can be registered.\n"
-"                    If no plugins were selected, registers any found plugin in the directories\n"
+"                    If no plugins were selected, any found plugin in the directories\n"
```
```suggestion
"                    If no plugins were registered, any found plugin in the directories\n"
```
according to the previous sentence.

@FedeDP

FedeDP replied:
You're right, nice catch :)

FedeDP committed:
Update userspace/sysdig/sysdig.cpp

Signed-off-by: Federico Di Pierro <nierro92@gmail.com>

Co-authored-by: Leonardo Grasso <me@leonardograsso.com>

FedeDP commented:
@leogr @LucaGuerra sorry, had to force push because i always forget how to commit github suggestion (you need to add the `Signed-off-by` and not the `Co-authored-by` ... damn) :angry:
